File: src/daybreak/adapters/terminal/obsidian.py
import logging
import os
from pathlib import Path

from daybreak.config import config

from .jsonc import dump_json_file, load_jsonc_file

logger = logging.getLogger("daybreak")

OBSIDIAN_GLOBAL_SETTINGS = Path(r"obsidian\obsidian.json")


class ObsidianAdapter:
    """Sets the "theme" key in Obsidian's global settings file
    (%APPDATA%\\obsidian\\obsidian.json), the app-wide default new vaults
    inherit.

    Deliberately does NOT also loop over every vault registered in that
    file and overwrite each vault's own .obsidian/app.json: that forces
    vaults with their own deliberate theme choice (e.g. a non-coding notes
    vault) to match Daybreak's mode too, which surprised a user in
    practice once this adapter was actually wired in. Global-only leaves
    per-vault choices alone.
    """

    name = "obsidian"

    def apply_mode(self, mode: str, theme_name: str, palette: dict = None):
        if os.name != "nt":
            return

        app_data = os.environ.get("APPDATA")
        if not app_data:
            return

        target_theme = config.get(
            "integrations",
            f"obsidian_{mode}_theme",
            "moonstone" if mode == "light" else "obsidian",
        )

        global_path = Path(app_data) / OBSIDIAN_GLOBAL_SETTINGS
        if not global_path.exists():
            logger.debug("Obsidian: global settings file not found, skipping.")
            return

        try:
            global_data = load_jsonc_file(global_path)
            if not isinstance(global_data, dict):
                return
            if global_data.get("theme") == target_theme:
                return
            global_data["theme"] = target_theme
            dump_json_file(global_path, global_data)
            logger.info(f"Obsidian: Applied theme '{target_theme}' to {global_path}")
        except Exception as exc:
            logger.warning(f"Obsidian: Failed to update {global_path}: {exc}")

File: go/internal/adapters/terminal/obsidian.go
//go:build windows

package terminal

import (
	"log"
	"os"
	"path/filepath"

	"daybreak/internal/config"
	"daybreak/internal/theme"
)

// ObsidianAdapter mirrors adapters/terminal/obsidian.py.
type ObsidianAdapter struct {
	Config *config.Manager
}

func (ObsidianAdapter) Name() string { return "obsidian" }

func (a ObsidianAdapter) ApplyMode(mode, _ string, _ theme.Palette) error {
	appData := os.Getenv("APPDATA")
	if appData == "" {
		return nil
	}

	def := "obsidian"
	if mode == "light" {
		def = "moonstone"
	}
	targetTheme := a.Config.GetIntegration("obsidian_"+mode+"_theme", def)

	globalPath := filepath.Join(appData, "obsidian", "obsidian.json")
	if _, err := os.Stat(globalPath); err != nil {
		return nil
	}

	var vaultPaths []string
	globalData, err := LoadJSONCFile(globalPath)
	if err != nil {
		log.Printf("Obsidian: Failed to update %s: %v", globalPath, err)
	} else {
		changed := false
		if cur, _ := globalData["theme"].(string); cur != targetTheme {
			globalData["theme"] = targetTheme
			changed = true
		}
		vaultPaths = extractVaultPaths(globalData)

		if changed {
			if err := DumpJSONFile(globalPath, globalData); err != nil {
				log.Printf("Obsidian: Failed to update %s: %v", globalPath, err)
			} else {
				log.Printf("Obsidian: Applied theme '%s' to %s", targetTheme, globalPath)
			}
		}
	}

	for _, vaultPath := range vaultPaths {
		appJSON := filepath.Join(vaultPath, ".obsidian", "app.json")
		if _, err := os.Stat(appJSON); err != nil {
			continue
		}
		appDataJSON, err := LoadJSONCFile(appJSON)
		if err != nil {
			log.Printf("Obsidian: Failed to update %s: %v", appJSON, err)
			continue
		}
		if cur, _ := appDataJSON["theme"].(string); cur == targetTheme {
			continue
		}
		appDataJSON["theme"] = targetTheme
		if err := DumpJSONFile(appJSON, appDataJSON); err != nil {
			log.Printf("Obsidian: Failed to update %s: %v", appJSON, err)
			continue
		}
		log.Printf("Obsidian: Applied theme '%s' to %s", targetTheme, appJSON)
	}

	return nil
}

func extractVaultPaths(globalData map[string]any) []string {
	vaults, ok := globalData["vaults"].(map[string]any)
	if !ok {
		return nil
	}

	var results []string
	for _, entry := range vaults {
		vaultEntry, ok := entry.(map[string]any)
		if !ok {
			continue
		}
		pathValue, ok := vaultEntry["path"].(string)
		if !ok || pathValue == "" {
			continue
		}
		if _, err := os.Stat(pathValue); err == nil {
			results = append(results, pathValue)
		}
	}
	return results
}
